src/commands/SinceCommand.ts

```diff
@@ -126,101 +74,78 @@ function formatResult(action: string, targetRoomId: string, recentJoins: Join[],
-    })
-    if ("error" in actionResult) {
-        return actionResult;
     }
-    const action: Action = actionResult.ok!;
+    if (!action) {
+        return {error: `Invalid <action>. Expected one of ${JSON.stringify(Action)}`};
+    }
 
-    // Now list affected rooms.
+    // Attempt to parse `<limit>` as a number.
+    const maxEntries = lexer.token("int").value as number;
+
+    // Parse rooms.
+    // Parse everything else as `<reason>`, stripping quotes if any have been added.
     const rooms: Set</* room id */string> = new Set();
-    let reasonParts: string[] | undefined;
-    for (let token of optionalTokens) {
-        const maybeArg = getTokenAsString(reasonParts ? "[reason]" : "[room]", token);
-        if ("error" in maybeArg) {
-            return maybeArg;
-        }
-        const maybeRoom = maybeArg.ok;
-        if (!reasonParts) {
-            // If we haven't reached the reason yet, attempt to use `maybeRoom` as a room.
-            if (maybeRoom === "*") {
-                for (let roomId of Object.keys(mjolnir.protectedRooms)) {
-                    rooms.add(roomId);
-                }
-                continue;
-            } else if (maybeRoom.startsWith("#") || maybeRoom.startsWith("!")) {
-                const roomId = await mjolnir.client.resolveRoom(maybeRoom);
-                if (!(roomId in mjolnir.protectedRooms)) {
-                    return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", `This room is not protected: ${htmlEscape(roomId)}.`);
-                }
+    let reason = "";
+    do {
+
+        let token = lexer.alternatives(
+            // Room
+            () => lexer.token("STAR"),
+            () => lexer.token("roomAliasOrID"),
+            // Reason
+            () => lexer.token("string"),
+            () => lexer.token("ETC")
+        );
+        if (!token || token.type === "EOF") {
+            // We have reached the end of rooms, no reason.
+            break;
+        } else if (token.type === "STAR") {
+            for (let roomId of Object.keys(mjolnir.protectedRooms)) {
                 rooms.add(roomId);
-                continue;
             }
-            // If we reach this step, it's not a room, so it must be a reason.
-            // All further arguments are now part of `reason`.
-            reasonParts = [];
+            continue;
+        } else if (token.type === "roomAliasOrID") {
+            const roomId = await mjolnir.client.resolveRoom(token.text);
+            if (!(roomId in mjolnir.protectedRooms)) {
+                return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", `This room is not protected: ${htmlEscape(roomId)}.`);
+            }
+            rooms.add(roomId);
+            continue;
+        } else if (token.type === "string" || token.type === "ETC") {
+            // We have reached the end of rooms with a reason.
+            reason = token.text;
+            break;
         }
-        reasonParts.push(maybeRoom);
-    }
-
+    } while(true);
```

Comment:
It doesn't seem right to do have to do this for each command

Reply:
In the longer term, I'd like to be able to specify a grammar for each command, but I'm not even sure of the shape of the DSL we'd need to specify such a grammar, so I'm planning to postpone this to a followup PR.

Reply:
The attempt is good and yes there is progress in that we have a table instead of a giant switch (supposedly).
There is also some reduced parsing overhead by having the defined tokens. 

However this is still incredibly hacky and maybe even less intuitive than before, so i don't think it's right to merge and follow up. 

```
describeCommand( {
    nameParts: string[], // e.g. ["rules", "matching"]
    arguments: Descriptions[], // Something we can generate the parsing from or just accept a callback to parse an argument instead
    cb: (commandRoomId: string, mjolnir: Mjolnir, event, ...arguments) => Promise<void>
}
```

The DSL doesn't have to be complicated I was imagining something like this but i haven't played around with it. I don't think it matters if we break some backwards incompatibilities since it probably is going to be worth the cost. Though maybe it is worth looking at some other bridges or bots to see what they do?

Reply:
Unfortunately, it gets complicated very quick. But I'll see if I can make something higher-level.

Reply:
Bots should not be communicating with Mjolnir via commands, not just because it's not meant for that but just because it'd be a really inefficient indirect way of achieving something (that is probably simple).

>Note that this is not sufficient to express subcommands such as config get/config set that have entirely different help messages, so I'd need to come up with something to handle these.

Well your example only shows arguments and not the whole command, but `nameParts` in the example above was meant for this situation e.g. `["config", "get"]` & `["config", "set"]`. Though that might not work.

I don't think the arguments should be expressed in terms of lexer tokens but what they actually should be e.g. we could say we want a user id but we probably won't want the raw string token but an instance of matrix-bot-sdk's `UserID` class (not sure this example is realistic enough, but you get the point that there is a difference). 
We will also probably will need to account for a situation where we want parse an argument ourselves using a callback (though with the lexer library this probably means parsing a token? rather than composing something from a token stream)